crates/interpreter/src/exec.rs

```diff
@@ -1043,8 +1050,15 @@ impl<'m> Thread<'m> {
         }
         let values = core::mem::take(self.values());
         let frame = self.frame();
-        let Label { arity, kind, .. } = frame.labels.drain(i ..).next().unwrap();
-        self.values().extend_from_slice(&values[values.len() - arity ..]);
+        let label_values_cnt_before: usize =
+            frame.labels[0 .. i].iter().map(|l| l.values_cnt).sum();
+        let label_values_cnt_after: usize =
+            frame.labels[i + 1 ..].iter().map(|l| l.values_cnt).sum();
+        let Label { arity, kind, values_cnt } = frame.labels.drain(i ..).next().unwrap();
+        self.values().extend_from_slice(&values[0 .. label_values_cnt_before]);
```

Comment:
This is extremely costly. We don't want to swap `values()`. We want to directly drain the proper range, which should be something like `label_values_cnt_before .. values.len() - label_values_cnt_after - values_cnt`.